Add "Auto Rotate" orientation for wide pages

timothyng-164 opened the pull request:
<!--
  Please include a summary of the change and which issue is fixed.
  Also make sure you've tested your code and also done a self-review of it.
  Don't forget to check all base themes and tablet mode for relevant changes.
  

  If your changes are visual, please provide images below:

### Images
| Image 1 | Image 2 |
| ------- | ------- |
| ![] | ![](https://github.githubassets.com/images/modules/logos_page/Octocat.png) |
-->

Closes #1977

This addresses [#1977](https://github.com/tachiyomiorg/tachiyomi/issues/1977) by adding the "Dual page rotate" option to the paged reader.

Feel free to edit and merge this branch as needed.

**New Reader Settings:**
- Dual page rotate
- Reverse dual page rotate orientation (only shown if the former is checked)

<a href="url"><img src="https://user-images.githubusercontent.com/25924483/190880507-a4aed82a-3750-4290-9f8d-003095f6f67d.png" align="center" height="30%" width="30%" ></a>
<hr></hr>

**Behavior for Dual Pages**:

1. Default Behavior for Dual Pages
2. "Dual page rotate" checked
3. "Reverse dual page rotate orientation" checked

<a href="url"><img src="https://user-images.githubusercontent.com/25924483/190880404-1b32aa6f-7ef2-4947-b585-e54a22fa9edd.png" align="left" height="25%" width="25%" ></a>

<a href="url"><img src="https://user-images.githubusercontent.com/25924483/190880391-08b95bcd-dcab-45c2-a338-b73e0f383990.png" align="left" height="25%" width="25%" ></a>

<a href="url"><img src="https://user-images.githubusercontent.com/25924483/190880409-00dca7cc-8c27-4d03-b72a-0314bfcfe655.png" align="left" height="25%" width="25%" ></a>

jferdous70 commented:
could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.

timothyng-164 commented:
> could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.

Sure, I'm thinking there could be 2 orientations:
- Auto Rotate Landscape - landscape on wide image, portrait on thin
- Auto Rotate Portrait - landscape on thin image, portrait on wide

I'm not sure if there's a better name but would that work?

jferdous70 commented:
> > could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.
> 
> Sure, I'm thinking there could be 2 orientations:
> 
>     * Auto Rotate Landscape - landscape on wide image, portrait on thin
> 
>     * Auto Rotate Portrait - landscape on thin image, portrait on wide
> 
> 
> I'm not sure if there's a better name but would that work?

That's not exactly what I wanted, but I now realized main tachiyomi don't about webtoon detection feature

timothyng-164 commented:
> > > could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.
> > 
> > 
> > Sure, I'm thinking there could be 2 orientations:
> > ```
> > * Auto Rotate Landscape - landscape on wide image, portrait on thin
> > 
> > * Auto Rotate Portrait - landscape on thin image, portrait on wide
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm not sure if there's a better name but would that work?
> 
> That's not exactly what I wanted, but I now realized main tachiyomi don't about webtoon detection feature

I see, I reverted th

> > > could auto rotate setting be made customizable? I would vastly prefer if it auto rotated to landscape for paged (manga) and portrait for long strip.
> > 
> > 
> > Sure, I'm thinking there could be 2 orientations:
> > ```
> > * Auto Rotate Landscape - landscape on wide image, portrait on thin
> > 
> > * Auto Rotate Portrait - landscape on thin image, portrait on wide
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > I'm not sure if there's a better name but would that work?
> 
> That's not exactly what I wanted, but I now realized main tachiyomi don't about webtoon detection feature

I see, I reverted the change to just have the "Auto Rotate" orientation".

arkon commented on app/src/main/java/eu/kanade/tachiyomi/ui/reader/ReaderActivity.kt:
```diff
@@ -778,6 +780,18 @@ class ReaderActivity : BaseRxActivity<ReaderPresenter>() {
+
+    private fun rotateOnWideImage(page: ReaderPage) {
+        val stream = page.stream!!
+        when (ImageUtil.isWideImage(stream.invoke())) {
```
You should only really need to read the header, similar to usages of the existing `isWideImage` function that takes in a buffered stream.

timothyng-164 replied:
I believe this has been addressed

arkon commented on app/src/main/java/eu/kanade/tachiyomi/util/system/ImageUtil.kt:
```diff
@@ -555,7 +565,7 @@ object ImageUtil {
      * Used to check an image's dimensions without loading it in the memory.
      */
-    private fun extractImageOptions(
+    fun extractImageOptions(
```
Can stay private?

timothyng-164 replied:
This has been reverted

arkon reviewed (commented):
General feedback:
- In its current implementation, I feel like this should only happen for paged reading modes. It's pretty janky to suddenly rotate in continuous vertical or webtoon modes.
- I'm wondering if we should just be loading a rotated version of the image rather than actually changing the system orientation, which requires a full activity recreation?

timothyng-164 commented:
Good point, I didn't consider continuous vertical or webtoons. 
I like the idea of just loading the image already rotated. I added a couple options to the reader settings.

arkon commented:
Rebased/cleaned up a bit in 953720472fe64ef488ecae7ae7fea453b8c7c68b
